vault/wrapping.go

```diff
@@ -0,0 +1,308 @@
+package vault
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/SermoDigital/jose/crypto"
+	"github.com/SermoDigital/jose/jws"
+	"github.com/SermoDigital/jose/jwt"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/vault/helper/jsonutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+const (
+	// The location of the key used to generate response-wrapping JWTs
+	coreWrappingJWTKeyPath = "core/wrapping/jwtkey"
+)
+
+func (c *Core) ensureWrappingKey() error {
+	entry, err := c.barrier.Get(coreWrappingJWTKeyPath)
+	if err != nil {
+		return err
+	}
+
+	var keyParams clusterKeyParams
+
+	if entry == nil {
+		key, err := ecdsa.GenerateKey(elliptic.P521(), rand.Reader)
+		if err != nil {
+			return errwrap.Wrapf("failed to generate wrapping key: {{err}}", err)
+		}
+		keyParams.D = key.D
+		keyParams.X = key.X
+		keyParams.Y = key.Y
+		keyParams.Type = corePrivateKeyTypeP521
+		val, err := jsonutil.EncodeJSON(keyParams)
+		if err != nil {
+			return errwrap.Wrapf("failed to encode wrapping key: {{err}}", err)
+		}
+		entry = &Entry{
+			Key:   coreWrappingJWTKeyPath,
+			Value: val,
+		}
+		if err = c.barrier.Put(entry); err != nil {
+			return errwrap.Wrapf("failed to store wrapping key: {{err}}", err)
+		}
+	}
+
+	// Redundant if we just created it, but in this case serves as a check anyways
+	if err = jsonutil.DecodeJSON(entry.Value, &keyParams); err != nil {
+		return errwrap.Wrapf("failed to decode wrapping key parameters: {{err}}", err)
+	}
+
+	c.wrappingJWTKey = &ecdsa.PrivateKey{
+		PublicKey: ecdsa.PublicKey{
+			Curve: elliptic.P521(),
+			X:     keyParams.X,
```

Comment:
Eh. Things will simply fail. Easy way to detect. There's really no reason it should ever happen and I'm not sure what to do in that situation anyways.